text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1611 @@
+  // Offset in the binary that corresponds to the first mapped address.
+  uint64 file_offset = 4;
+  // The object this entry is loaded from.  This can be a filename on
+  // disk for the main binary and shared libraries, or virtual
+  // abstractions like "[vdso]".
+  int64 filename = 5;  // Index into string table
+  // A string that uniquely identifies a particular program version
+  // with high probability. E.g., for binaries generated by GNU tools,
+  // it could be the contents of the .note.gnu.build-id field.
+  int64 build_id = 6;  // Index into string table
+  // Specifies the kind of build id. See BuildIdKind enum for more details [optional]
+  BuildIdKind build_id_kind = 11;
+  // References to attributes in Profile.attribute_table. [optional]
+  repeated uint64 attributes = 12;
+  // The following fields indicate the resolution of symbolic info.
+  bool has_functions = 7;
+  bool has_filenames = 8;
+  bool has_line_numbers = 9;
+  bool has_inline_frames = 10;
+}
+
+// Describes function and line table debug information.
+message Location {
+  // Unique nonzero id for the location.  A profile could use
+  // instruction addresses or any integer sequence as ids. [deprecated]
+  uint64 id = 1;
+  // The index of the corresponding profile.Mapping for this location.
+  // It can be unset if the mapping is unknown or not applicable for
+  // this profile type.
+  uint64 mapping_index = 2;
+  // The instruction address for this location, if available.  It
+  // should be within [Mapping.memory_start...Mapping.memory_limit]
+  // for the corresponding mapping. A non-leaf address may be in the
+  // middle of a call instruction. It is up to display tools to find
+  // the beginning of the instruction if necessary.
+  uint64 address = 3;
+  // Multiple line indicates this location has inlined functions,
+  // where the last entry represents the caller into which the
+  // preceding entries were inlined.
+  //
+  // E.g., if memcpy() is inlined into printf:
+  //    line[0].function_name == "memcpy"
+  //    line[1].function_name == "printf"
+  repeated Line line = 4;
+  // Provides an indication that multiple symbols map to this location's
+  // address, for example due to identical code folding by the linker. In that
+  // case the line information above represents one of the multiple
+  // symbols. This field must be recomputed when the symbolization state of the
+  // profile changes.
+  bool is_folded = 5;
+
+  // Type of frame (e.g. kernel, native, python, hotspot, php). Index into string table.
+  uint32 type_index = 6;
+
+  // References to attributes in Profile.attribute_table. [optional]
+  repeated uint64 attributes = 7;
+}
+
+// Details a specific line in a source code, linked to a function.
+message Line {
+  // The index of the corresponding profile.Function for this line.
+  uint64 function_index = 1;
+  // Line number in source code.
+  int64 line = 2;
+  // Column number in source code.
+  int64 column = 3;
```

Comment:
Python 3.11 added column-level information with [this PEP](https://peps.python.org/pep-0657/). The motivation is primarily for the tracebacks displayed when the application ends with an uncaught exception, but it is applicable to the broader spectrum of debugging tools (including profilers). The enhanced location information comes in the form of line and column ranges. That is, Python code objects have a lookup table that can map bytecode offsets to line start/end and column start/end pairs.

As some of the [critique](https://lists.archive.carbon60.com/python/dev/1531870) has pointed out, this new location information poses a problem for data "correctness". Because the PEP introduced ranges, simply extracting the line start to fill in the line field of the pprof spec could lead to potentially incorrect location information. For example, a profiler that is extracting/receiving only the line start information will not account for the wall time (or whatever other metric is being measured) for the full line range, thus introducing a form of bias that could be misleading when the data is (visually) represented (e.g. source code heat maps highlighting just the first line of the range).

Based on the points raised above, I would like to propose extending this structure to allow carrying the information of line end and column end, when this information is available from the runtime.

Reply:
This was discussed [here](https://github.com/google/pprof/pull/818#issuecomment-1896267978) a bit. I definitely wonder what others think but in my opinion expressing line or column ranges is fairly niche and should not be a part of the initial revision. I'd rather discuss this as a separate change request later, collecting use cases, pros and cons etc.

Reply:
I agree that I don't think this has to hold up this OTEP. But let's try to take another look at it as part of the upcoming spec work.